### DISCUSSION TITLE
Manifold constructor/union gives wrong results

### DISCUSSION THREAD

**Discussion by johnnybigert** [elalish/manifold#487] [Q&A]
Hi!

I'm using the Manifold constructor to union a lot of slightly overlapping geometries. Over time, I have noticed that I often get strange results in the form of "antennas" (thin wedges) inside of the solid, and now I finally got some time to open a discussion about it here :)

In the example I provide below, I have a list of 200+ geometries that I pass to the Manifold constructor. By binary searching, I found that strange things happened when I added the 183:rd geometry (that is, having 182 geometries in the list and running the constructor was fine, while adding the 183:rd to the list and running the constructor resulted in strange artifacts in the mesh). So obviously, I suspected there was something wrong with the 183:rd, but it looks fine and is very similar to all other shapes.

Interestingly, the strange artifacts that appear in the mesh are nowhere near the bounding box of shape 183! Furthermore, if I remove some of the other geometries (before 183), adding 183 produces no artifacts. So the issue is complex and it's very hard to produce a minimal example that reproduces the problem. If I remove 183 from the list, artifacts are still created from other geometries.

Before adding geometry 183 (which is shown in red) to the list that is passed to the Manifold constructor:
<img width="595" alt="Screenshot 2023-06-26 at 15 27 26" src="https://github.com/elalish/manifold/assets/17929792/2768dbcd-1823-43aa-8238-c2d3ea045e66">

After adding geometry 183 (red) to the list that is passed to the Manifold constructor. The artifacts (red arrows) now appear, and they are nowhere near 183.
<img width="653" alt="Screenshot 2023-06-26 at 15 28 12" src="https://github.com/elalish/manifold/assets/17929792/2766f269-4c1f-4736-a1d1-e82e1ad8bcdc">

Geometry 183 is the extrude of the following CrossSection shape. It's counter-clockwise and simple.
<img width="543" alt="Screenshot 2023-06-26 at 15 28 32" src="https://github.com/elalish/manifold/assets/17929792/5e06af7c-e1e1-4058-9f92-a95dd43c1f7f">

My question is: How do I best help you reproduce the problem? Do you have an OFF importer in C++? If so, I could provide you with a C++ example reproducing the problem.

**Comment by geoffder** [comment]
What happens if you scale up everything first?

**Comment by elalish** [comment]
You keep talking about passing an array of objects to the Manifold constructor, but I don't think we have a constructor that takes an array of geometries. Are you doing Boolean operations to combine these? Or `Compose`? 

And yes, `meshIO` uses Assimp to import a variety of formats, and I *think* it supports OFF. It would be very helpful to see your repro code.

**Comment by johnnybigert** [comment]
> You keep talking about passing an array of objects to the Manifold constructor, but I don't think we have a constructor that takes an array of geometries. Are you doing Boolean operations to combine these? Or `Compose`?
> 
> And yes, `meshIO` uses Assimp to import a variety of formats, and I _think_ it supports OFF. It would be very helpful to see your repro code.

Sorry, I should have been clear that I'm using the Python bindings :) And the Python code is just a `for` loop doing `first += manifolds[i]`.

I will try to find some time to reproduce it in C++.

**Comment by elalish** [comment]
Ah, so it's not a constructor at all, but sequential Union operations. Do the artifacts have any volume, or only surface area? Anything you can do to minimize the repro would be much appreciated. It looks like the upper surface has a lot of nearly coplanar geometry - is that important? You might consider running `AsOriginal()` an an occasional intermediate step to clear that kind of thing out.

**Comment by johnnybigert** [comment]
@elalish They all have non-zero volume.
The co-planar surfaces are important, but I'm only interested in them being co-planar within a threshold of eps=0.0001 or so. Any deviance under that will be of no consequence to subsequent computations.

I will try to see what happens if I scale things up as @geoffder suggested. And after that, when I get the time, I will try to reproduce in C++.

**Comment by johnnybigert** [comment]
Scaling up made the matter worse. I tried scaling all values down to abs(value) < 1.0 and it worked like a charm! Huge thanks for the work-around!
My values was typically between -100 and +141, so nothing extreme.

**Comment by elalish** [comment]
Glad it's working, but I'm still interested in getting a repro of this. What about the co-planar surfaces is important? I should have been more specific - we will avoid merging those triangles together into a larger face because we keep the identities of the meshes they came from separate. If you aren't re-applying mesh properties, then `AsOriginal` is a good thing to call to allow us to merge them and simplify the geometry for further operations.

**Reply by elalish** [reply]
I'm moving this to a discussion since we don't have an actionable repro yet.
